src/databasetasks.cpp

```diff
@@ -85,6 +85,13 @@ void DatabaseTasks::runTask(const DatabaseTask& task)
 
 void DatabaseTasks::flush()
 {
+	taskLock.lock();
+	flushInternal();
+	taskLock.unlock();
+}
+
+void DatabaseTasks::flushInternal()
+{
 	while (!tasks.empty()) {
 		runTask(tasks.front());
```

Comment:
This may deadlock if the callback attempts to query another async database operation.